// my-v0-project/src/pages/SiapaDisana/SiapaDisana.jsx
import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { FiArrowLeft } from 'react-icons/fi';
import { useReport } from '../../context/ReportContext'; // ← TAMBAH INI
import './SiapaDisana.css';
import logoImage from '../../assets/logo pojok kanan .png';
import bearImage from '../../assets/sapa.png';
import { reportStorage } from '../../utils/reportStorage';
import { createChat, addMessage } from '../../utils/chatStorage';

function SiapaDisana() {
  const navigate = useNavigate();
  const { reportData, updateMultipleFields, submitReport } = useReport(); // ← TAMBAH INI
  const [selectedPerson, setSelectedPerson] = useState('');
  const [customPerson, setCustomPerson] = useState('');
  const [loading, setLoading] = useState(false); // ← TAMBAH INI

  const peopleOptions = [
    { icon: '👥', label: 'Teman sekelasku' },
    { icon: '👤', label: 'Kakak sekelasku' },
    { icon: '👨‍🏫', label: 'Guru / wali kelas melihat' },
    { icon: '👁️', label: 'Orang dewasa disekolah lihat' },
    { icon: '📱', label: 'Teman di HP / Chat' }
  ];

  const playSound = () => {
    const audioContext = new (window.AudioContext || window.webkitAudioContext)();
    const oscillator = audioContext.createOscillator();
    const gainNode = audioContext.createGain();

    oscillator.connect(gainNode);
    gainNode.connect(audioContext.destination);

    oscillator.frequency.value = 600;
    oscillator.type = 'sine';

    gainNode.gain.setValueAtTime(0.3, audioContext.currentTime);
    gainNode.gain.exponentialRampToValueAtTime(0.01, audioContext.currentTime + 0.2);

    oscillator.start(audioContext.currentTime);
    oscillator.stop(audioContext.currentTime + 0.2);
  };

  const handleBackClick = () => {
    playSound();
    setTimeout(() => {
      navigate(-1);
    }, 200);
  };

  const handlePersonClick = (person) => {
    setSelectedPerson(person);
    setCustomPerson('');
    playSound();
  };

  const handleCustomPersonChange = (e) => {
    setCustomPerson(e.target.value);
    setSelectedPerson('');
  };

  const handleSubmit = async () => { // ← UBAH JADI ASYNC
    const finalPerson = customPerson.trim() || selectedPerson;
    
    if (!finalPerson) {
      alert('Mohon pilih atau tulis siapa yang ada disana!');
      return;
    }

    playSound();
    setLoading(true); // ← TAMBAH INI
    
    try {
      // Update data saksi terakhir
      updateMultipleFields({
        witnesses: [finalPerson],
        witnessesOther: customPerson.trim()
      });
      
      // Ambil semua data dari Context (sudah dikumpulkan dari halaman sebelumnya)
      const bullyingType = reportData.type || sessionStorage.getItem('bullyingType') || 'Tidak disebutkan';
      const storyDetail = reportData.description || sessionStorage.getItem('storyDetail') || '';
      const tanggalKejadian = reportData.date || sessionStorage.getItem('tanggalKejadian') || '';
      const lokasiKejadian = reportData.location || sessionStorage.getItem('lokasiKejadian') || '';
      
      // Format tanggal
      const dateObj = new Date(tanggalKejadian);
      const formattedDate = dateObj.toLocaleDateString('id-ID', {
        day: '2-digit',
        month: '2-digit',
        year: 'numeric'
      });
      
      console.log('📦 Data yang akan dikirim ke database:', {
        type: bullyingType,
        description: storyDetail,
        date: tanggalKejadian,
        location: lokasiKejadian,
        witnesses: [finalPerson]
      });
      
      // ========== KIRIM KE DATABASE MONGODB ========== 
      const result = await submitReport();
      console.log('✅ Berhasil disimpan ke database:', result);
      
      // ========== SIMPAN KE LOCALSTORAGE (untuk fitur lokal) ==========
      const newReport = reportStorage.saveReport({
        type: bullyingType,
        detail: storyDetail,
        tanggal: formattedDate,
        tempat: lokasiKejadian,
        siapa: finalPerson,
        date: formattedDate,
        details: {
          what: storyDetail,
          when: formattedDate,
          where: lokasiKejadian,
          who: finalPerson
        }
      });
      
      // Buat chat session dengan kode = report ID dari MongoDB
      const chatCode = result._id || newReport.id.toString();
      createChat(chatCode, bullyingType);
      
      // Tambahkan pesan pertama dari siswa
      if (storyDetail.trim()) {
        addMessage(chatCode, storyDetail, 'student');
      }
      
      // Bersihkan sessionStorage
      sessionStorage.removeItem('bullyingType');
      sessionStorage.removeItem('storyDetail');
      sessionStorage.removeItem('tanggalKejadian');
      sessionStorage.removeItem('lokasiKejadian');
      
      // Simpan report ID dari MongoDB untuk halaman sukses
      sessionStorage.setItem('reportId', result._id);
      
      setTimeout(() => {
        navigate('/laporan-success', {
          state: { reportId: result._id } // ← KIRIM ID KE HALAMAN SUCCESS
        });
      }, 200);
      
    } catch (error) {
      console.error('❌ Error menyimpan laporan:', error);
      alert('Gagal mengirim laporan: ' + error.message + '\n\nSilakan coba lagi.');
      setLoading(false);
    }
  };

  return (
    <div className="siapa-disana-page">
      {/* Header */}
      <div className="page-header">
        <button className="back-button" onClick={handleBackClick}>
          <FiArrowLeft className="back-icon" />
          <span>BACK</span>
        </button>

        <div className="header-title">
          <h1>SAHABAT DIGITAL ANTI BULLYING</h1>
        </div>

        <div className="corner-logo">
          <img src={logoImage} alt="Safe School Logo" />
        </div>
      </div>

      {/* Content Section */}
      <div className="siapa-content">
        {/* Bear and Speech Bubble */}
        <div className="bear-section">
          <div className="speech-bubble">
            <p>Sekarang coba pilih ya, siapa aja yang ada waktu itu?</p>
          </div>
          <img src={bearImage} alt="Bear Character" className="bear-character" />
        </div>

        {/* Form Section */}
        <div className="form-section">
          <div className="question-bubble">
            <p>siapa yang ada disana? 🤔</p>
          </div>

          {/* People Options */}
          <div className="people-buttons">
            {peopleOptions.map((person, index) => (
              <button
                key={index}
                className={`person-btn ${selectedPerson === person.label ? 'active' : ''}`}
                onClick={() => handlePersonClick(person.label)}
                disabled={loading}
              >
                <span className="person-icon">{person.icon}</span>
                <span className="person-label">{person.label}</span>
              </button>
            ))}
          </div>

          {/* Custom Input */}
          <div className="custom-person-section">
            <div className="speech-bubble-small">
              <p>isi disini jika tidak ada dipilihan ya...</p>
            </div>
            <input 
              type="text" 
              value={customPerson}
              onChange={handleCustomPersonChange}
              className="custom-person-input"
              placeholder="Tulis siapa yang ada disana..."
              disabled={loading}
            />
          </div>

          {/* Submit Button */}
          <button 
            className="submit-button" 
            onClick={handleSubmit}
            disabled={loading}
          >
            <span>{loading ? 'Mengirim...' : 'Kirim'}</span>
          </button>
        </div>
      </div>
    </div>
  );
}

export default SiapaDisana;
